File: 백준/Silver/1182. 부분수열의 합/부분수열의 합.cc
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

int N, S;
long long ans;
vector<int> v;
vector<bool> visited;

void Input() {
   
    cin >> N >> S;
    for (int i = 0; i < N; i++) {
        int num;
        cin >> num;
        v.push_back(num);
        visited.push_back(false);
    }

}

void DFS(int n) {

    if (n == N) {
        long long sum = 0;

        for (int i = 0; i < N; i++) {
            if (visited[i]) {
                sum += v[i];
            }
        }

        if (sum == S)
            ans++;
        return;
    }

    visited[n] = true;
    DFS(n + 1);

    visited[n] = false;
    DFS(n + 1);
}



void Solve() {
    ans = 0; 
    DFS(0);

    if (S == 0)
        ans--;
    cout << ans << "\n";
}


int main() {
    ios::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);

    Input();
    Solve();

    return 0;
}

File: 백준/Silver/1182. 부분수열의 합/부분수열의 합.java
import java.util.*;
import java.io.*;



public class Main {

    static int N, S;
    static int[] arr;
    static boolean[] visited;
    static int ans = 0;

    static void DFS(int n, int sum){
        
        if(n == N){
            if(sum == S) ans++;
            return;
        }
        DFS(n + 1, sum + arr[n]);
        DFS(n + 1, sum);
    }

    static void Input() throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        S = Integer.parseInt(st.nextToken());
        arr = new int[N];
        visited = new boolean[N];
        st = new StringTokenizer(br.readLine());
        for(int i = 0; i < N; i++){ 
            arr[i] = Integer.parseInt(st.nextToken());
        }
    }

    public static void main(String[] args) throws IOException{
        
        Input();
        DFS(0, 0);
        if(S == 0)  ans--;
        System.out.println(ans);
    }
}
